docs/source/dev/index.rst

```diff
@@ -43,49 +43,141 @@ For other questions regarding OpenFAST, please contact
-documentation which is compiled with Doxygen. Though this portion of the
-documentation is always under development, the existing API reference can
-be found in the following pages:
-
-- `Main Page <../../html/index.html>`_
-- `Index of Types <../../html/classes.html>`_
-- `Source Files <../../html/files.html>`_
-
 .. _development_philosophy:
 
-Development Philosophy
-~~~~~~~~~~~~~~~~~~~~~~
-
-OpenFAST is intended to be a self sustaining community developed software.
-A couple of tenets of this goal are that the code should be reasonably
-straightforward to comprehend and manageable to improve. With that in mind, we
-expect that new capabilities will include adequate testing and documentation.
-
-We have the following guidance for developers:
-
-- When fixing a bug, first introduce a unit test that exposes the bug, fix the
-  bug, and submit a Pull Request. See :ref:`testing` and
-  :ref:`github_workflow` for more information.
-
-- When adding a new feature, create appropriate automated unit and regression
-  tests as described in :ref:`testing`. The objective is to create a GitHub
-  pull request that provides adequate verification and validation so that the
-  NREL OpenFAST developer team can merge the pull request with confidence that
-  the new feature is "correct" and supports our goal of self-sustaining
-  software. See :ref:`pull_requests` for more information on submitting
-  a pull request.
-
-- If a code modification affects regression test results in an expected manner,
-  work with the NREL OpenFAST developer team to upgrade the regression test
-  suite via a GitHub issue or pull request at the `openfast/r-test <https://github.com/openfast/r-test>`_
-  repository.
-
-Development Guidelines
-~~~~~~~~~~~~~~~~~~~~~~
-The following sections provide extended guidance on how to develop source code,
+Development Philosophy and Guidelines
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+OpenFAST is intended to be a self-sustaining, community developed software.
+While the NREL OpenFAST team serves as the gatekeeper of the repository, we
+actively encourage the community to share new ideas and contribute code.
+Considerations for contributing code are outlined here.
+
+Engagement with NREL
+--------------------
+
+The process for community code contribution starts with engaging directly
+with the NREL OpenFAST team to define the scope of the work and coordinate
+development efforts. This is particularly important since many groups
+work on OpenFAST simultaneously. By engaging early, all developers can
+stay up to date and minimize conflicts during the code merge.
+The prefered method of communication is `GitHub Issues <https://github.com/openfast/openfast/issues>`_.
+An initial post should contain all relevant information about the planned
+development work, the areas of the software that will be impacted,
+and any model validation materials. After this step, the NREL OpenFAST
+team may reach out to schedule a meeting to talk through the details.
```

Comment:
For significant code contributions or feature enhancements, I would recommend that developers write and share an implementation plan before modifying the source code.  It is far easier to review, iterate, and agree on the plan before source code is written.  An implementation plan will greatly aid in the programming effort and is a useful jumping off point for writing the module’s user’s guide and theory manual.

A good plan within the modularization framework of OpenFAST will follow the definitions and nomenclature used by the NWTC Programmer's Handbook: http://www.nrel.gov/docs/fy13osti/57228.pdf, and should clearly:

- State whether the module is intended for loose coupling, tight coupling for time marching, and/or linearization.
- Define the module’s inputs (including initialization), outputs (including initialization), states (continous, discrete, and constraint), and parameters, including units.
- Lay out an example input file for the module.
- Explain the module’s mathemetical formulation, including Jacobians (for tight coupling and linearization), in the form required of the framework.
- Prescribe how the module’s inputs are derived from the outputs of other specific modules.
- Identify any potential numerical problems and how to avoid them in the code.
- Lay out the module’s subroutines using pseudocode (as opposed to actual code), including identifying which mathematical formulas are used by which subroutines, and describing the algorithms used in the solution process.

There are many good examples of implementation plans now shared through readthedocs.

Reply:
Thanks for the suggestion, I've added a section for this with your comments above and links to other implementation plans